npep/npep-126-egress-traffic-control.md

```diff
@@ -76,13 +76,249 @@ selected cluster workloads to k8s-apiservers for securing the server.
+    to:
+    - nodes:
+        matchLabels:
+          node-role.kubernetes.io/control-plane: true
+    ports:
+      - portNumber:
+          protocol: TCP
+          port: 443
+      - portNumber:
+          protocol: TCP
+          port: 6443
+```
+
+### Implementing egress traffic control towards external destinations
+
+This NPEP proposes to add a new type of `AdminNetworkPolicyEgressPeer` called `ExternalNetworks`
+to be able to explicitly select external destinations (based on the `externalNetworkSet`'s
+labels) in the cluster. This peer type will not be supported in `AdminNetworkPolicyIngressPeer`.
+
+```
+// AdminNetworkPolicyEgressPeer defines a peer to allow traffic to.
+// Exactly one of the selector pointers must be set for a given peer. If a
+// consumer observes none of its fields are set, they must assume an unknown
+// option has been specified and fail closed.
+// +kubebuilder:validation:MaxProperties=1
+// +kubebuilder:validation:MinProperties=1
+type AdminNetworkPolicyEgressPeer struct {
+    <snipped>
+	// ExternalNetworks defines a way to select ExternalNetworkSets
+	// that consist of network CIDRs that live outside the cluster as a peer.
+	// This field follows standard label selector semantics; if present
+	// but empty, it selects all ExternalNetworkSets defined in the cluster.
+	//
+	// Support: Core
+	//
+	// +optional
+	ExternalNetworks *metav1.LabelSelector `json:"externalNetworks,omitempty"`
+}
+```
+
+An `externalNetworkSet` is a new object used to define a set of networks outside
+the cluster. This is defined as a new object so that its easy for users to group
+their external entities and then refer to that group from the different egress rules.
+This ensures consistency without the users having to manually change the ANP/BANP
+to accommodate the changing needs of the network cidrs. They can directly edit the
+new CRD which guarantees all rules effecting that peer will take effect automatically.
+In the future when we define `DeveloperNetworkPolicies`, this will come even more handy
+for uniformity specially across `Pass` actions.
+
+```
+// ExternalNetworkSet is a cluster level resource that is used to define
+// a set of networks outsides the cluster which can be referred to from
+// the AdminNetworkPolicy && BaselineAdminNetworkPolicy APIs as an external peer
+type ExternalNetworkSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Specification of the desired behavior of ExternalNetworkSet.
+	Spec ExternalNetworkSetSpec `json:"spec"`
+}
+
+// ExternalNetworkSetSpec defines the desired state of ExternalNetworkSet.
+type ExternalNetworkSetSpec struct {
+	// Networks is the list of NetworkCIDR (both v4 & v6) that can be used to define
+	// external destinations.
+	// A total of 100 CIDRs will be allowed in each NetworkSet instance.
+	// ANP & BANP APIs may use the .spec.in(e)gress.from(to).externalNetworks selector
+	// to select a set of external networks
```

Comment:
you said it was egress-only above

Reply:
oh oops yes, earlier I had a CEL expression and I changed this to splitting the ingress and egress as two different structs, but this got left over. Editing this to be egress only. nice catch.